integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/search/reference/FieldReferenceIT.java

```diff
@@ -0,0 +1,524 @@
+								.asList()
+				)
+				.where( f -> f.matchAll() ) )
+				.hasHitsAnyOrder(
+						List.of( "string_n_0", SomeEnum.VALUE_1 ),
+						List.of( "string_n_1", SomeEnum.VALUE_2 ),
+						List.of( "string_n_2", SomeEnum.VALUE_3 )
+				);
+
+		assertThatQuery( index.query()
+				.select(
+						f -> f.field( IndexedEntity_.vector )
+				)
+				.where( f -> f.knn( 10 ).field( IndexedEntity_.vector )
+						.matching( new float[] { 1.0f, 1.0f, 1.0f } ) ) )
+				.hasHitsAnyOrder(
+						new float[] { 0.0f, 0.0f, 0.0f },
+						new float[] { 10.0f, 100.0f, 1000.0f },
+						new float[] { 20.0f, 200.0f, 2000.0f }
+				);
+
+		assertThatQuery( index.query()
+				.where( f -> f.nested( IndexedEntity_.nestedEmbeddedEntity )
+						.add( f.match().field( IndexedEntity_.nestedEmbeddedEntity.someEnum )
+								.matching( SomeEnum.VALUE_1 ) ) ) )
+				.hasDocRefHitsAnyOrder( index.typeName(), dataSet.docId( 0 ) );
+
+		assertThatQuery( index.query()
+				.select( f -> f.field( IndexedEntity_.stringProjectedAsBytes ) )
+				.where( f -> f.matchAll() )
+		).hasHitsAnyOrder(
+				new byte[] { 97, 95, 48 },
+				new byte[] { 97, 95, 50 },
+				new byte[] { 97, 95, 49 }
+		);
+
+		assertThatQuery( index.query()
+				.select( f -> f.field( IndexedEntity_.stringProjectedAsBytes.noConverter() ) )
+				.where( f -> f.matchAll() )
+		).hasHitsAnyOrder(
+				"a_0", "a_1", "a_2"
+		);
+	}
+
+	private static class IndexedEntity {
+		private int primitiveInteger;
+		private Double wrapperDouble;
+		private String string;
+		private String stringProjectedAsBytes;
+		private LocalDate localDate;
+		private float[] vector;
+
+		private List<String> stringList;
+
+		private EmbeddedEntity nestedEmbeddedEntity;
+		private EmbeddedEntity flattenedEmbeddedEntity;
+
+		private List<EmbeddedEntity> embeddedEntityList;
+
+		// @IndexedEmbedded(includeDepth = 2,
+		// 		structure = ObjectStructure.NESTED,
+		// 		excludePaths = {
+		// 				"parent.parent.primitiveInteger"
+		// 		// ... (other paths so that only string and vector fields are included)
+		// 		}
+		// )
+		private IndexedEntity parent;
+
+	}
+
+	private static class EmbeddedEntity {
+		private String string;
+		private SomeEnum someEnum;
+	}
+
+	private enum SomeEnum {
+		VALUE_1, VALUE_2, VALUE_3
+	}
+
+	public static class IndexedEntity_ {
+		public static ValueFieldReference<Integer, Integer, Integer> primitiveInteger;
+		public static ValueFieldReference<Double, Double, Double> wrapperDouble;
+		public static ValueFieldReference<String, String, String> string;
+		public static ValueFieldReference<String, String, byte[]> stringProjectedAsBytes;
+		public static ValueFieldReference<LocalDate, LocalDate, LocalDate> localDate;
+		public static ValueFieldReference<float[], float[], float[]> vector;
+
+		public static ValueFieldReference<String, String, String> stringList;
+		public static NestedEmbeddedEntity_ nestedEmbeddedEntity;
+		public static EmbeddedEntityList_ embeddedEntityList;
```

Comment:
If I got your idea correctly here, instead of working with `EmbeddedEntity` we are generating a type with references for each of the object fields depending on what will actually end up in them.